### PR TITLE
Implement per persistence unit configuration for Hibernate Envers

### LINKED ISSUE #28133: Implement per persistence unit configuration for Hibernate Envers
Hibernate Envers actually has the existing infrastructure to do so as we implement the `active` boolean per persistence unit but all the rest of the config is still global.
We should move everything to the persistence unit specific config, adjust `HibernateEnversRecorder#contributeBootProperties()` to target per PU config for all properties and it would probably work. If we're lucky.

We also need a proper test for that new feature.

Typically, we want to be able to do:

```
quarkus.hibernate-envers.db1.store-data-at-delete=true
quarkus.hibernate-envers.db1.audit-table-suffix=_aud
quarkus.hibernate-envers.db1.revision-field-name=rev
quarkus.hibernate-envers.db1.revision-type-field-name=revtype
quarkus.hibernate-envers.db1.audit-strategy=org.hibernate.envers.strategy.ValidityAuditStrategy
quarkus.hibernate-envers.db1.audit-strategy-validity-end-rev-field-name=revend
quarkus.hibernate-envers.db1.audit-strategy-validity-store-revend-timestamp=true
quarkus.hibernate-envers.db1.audit-strategy-validity-revend-timestamp-field-name=revend_tstmp

quarkus.hibernate-envers.db2.audit-table-suffix=_aud
quarkus.hibernate-envers.db2.revision-field-name=revision_id
quarkus.hibernate-envers.db2.revision-type-field-name=revtype
quarkus.hibernate-envers.db2.do-not-audit-optimistic-locking-field=false
quarkus.hibernate-envers.db2.store-data-at-delete=true
```

### Discussed in https://github.com/quarkusio/quarkus/discussions/27971

<div type='discussions-op-text'>

<sup>Originally posted by **Traivor** September 15, 2022</sup>
It doesn't appear that the hibernate-envers plugin supports configuration on a persistence unit basis. Am I reading things right? Are there any plans to support this?

Alternatively, would passing envers properties via the new quarkus.hibernate-orm."unit-name".unsupported-properties property work?</div>

### PR DESCRIPTION
**Pull request by tmihalac**
Fixes https://github.com/quarkusio/quarkus/issues/28133
Signed-off-by: Theodor Mihalache <tmihalac@redhat.com>

### CODE REVIEW

File: extensions/hibernate-envers/deployment/src/main/java/io/quarkus/hibernate/envers/deployment/HibernateEnversProcessor.java [hunk 1]
```diff
@@ -38,8 +40,14 @@ public void registerEnversReflections(BuildProducer<ReflectiveClassBuildItem> re
...
-        buildTimeConfig.revisionListener.ifPresent(s -> reflectiveClass.produce(new ReflectiveClassBuildItem(true, true, s)));
-        buildTimeConfig.auditStrategy.ifPresent(s -> reflectiveClass.produce(new ReflectiveClassBuildItem(true, true, s)));
+        final Map<String, HibernateEnversBuildTimeConfigPersistenceUnit> allPersistenceUnitConfigsAsMap = buildTimeConfig
+                .getAllPersistenceUnitConfigsAsMap();
+        if (allPersistenceUnitConfigsAsMap != null && !allPersistenceUnitConfigsAsMap.isEmpty()) {
+            for (HibernateEnversBuildTimeConfigPersistenceUnit pu : allPersistenceUnitConfigsAsMap.values()) {
+                pu.revisionListener.ifPresent(s -> reflectiveClass.produce(new ReflectiveClassBuildItem(true, true, s)));
+                pu.auditStrategy.ifPresent(s -> reflectiveClass.produce(new ReflectiveClassBuildItem(true, true, s)));
+            }
+        }
```
Comment (yrodiere) [inline extensions/hibernate-envers/deployment/src/main/java/io/quarkus/hibernate/envers/deployment/HibernateEnversProcessor.java:50]: No need to be *that* defensive: `getAllPersistenceUnitConfigsAsMap` never returns null, and if it's empty the loop just won't do anything:

```suggestion
        for (HibernateEnversBuildTimeConfigPersistenceUnit pu : buildTimeConfig.getAllPersistenceUnitConfigsAsMap().values()) {
            pu.revisionListener.ifPresent(s -> reflectiveClass.produce(new ReflectiveClassBuildItem(true, true, s)));
            pu.auditStrategy.ifPresent(s -> reflectiveClass.produce(new ReflectiveClassBuildItem(true, true, s)));
        }
```

File: extensions/hibernate-envers/deployment/src/test/java/io/quarkus/hibernate/orm/envers/EnversConfigurationPerPUTest.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        assertThat(configuredRevisionTypeName).isEqualTo("REV_TYPE");
+
+        configuredRevisionTypeName = getAuditConfiguration("db2").getRevisionTypePropName();
+        assertThat(configuredRevisionTypeName).isEqualTo("REVTYPE");
+    }
+
+    private AuditEntitiesConfiguration getAuditConfiguration(String auditConfig) {
+        switch (auditConfig) {
+            case "db1":
+                return ((((SessionFactoryImplementor) emf1
```
Comment (yrodiere) [inline extensions/hibernate-envers/deployment/src/test/java/io/quarkus/hibernate/orm/envers/EnversConfigurationPerPUTest.java:76]: Just a nitpick, no need to change anything, but you could have just passed the `EntityManagerFactory` as an argument to this method directly.

File: extensions/hibernate-envers/runtime/src/main/java/io/quarkus/hibernate/envers/HibernateEnversRecorder.java [hunk 3]
```diff
@@ -32,43 +32,43 @@ private HibernateEnversIntegrationStaticInitListener(HibernateEnversBuildTimeCon
         @Override
         public void contributeBootProperties(BiConsumer<String, Object> propertyCollector) {
             var puConfig = buildTimeConfig.getAllPersistenceUnitConfigsAsMap().get(puName);
-            if (puConfig != null && puConfig.active.isPresent() && !puConfig.active.get()) {
+            if (puConfig == null || (puConfig.active.isPresent() && !puConfig.active.get())) {
                 propertyCollector.accept(EnversService.INTEGRATION_ENABLED, "false");
                 // Do not process other properties: Hibernate Envers is inactive anyway.
                 return;
             }
```
Comment (yrodiere) [inline extensions/hibernate-envers/runtime/src/main/java/io/quarkus/hibernate/envers/HibernateEnversRecorder.java:39]: We generally take the approach that extensions are activated automatically by default, as soon as they're in the classpath. That is, `active` is intended to default to `true`. So... why this change in behavior here?

If you just wanted to avoid a `NullPointerException`, I think this would be more correct:

```suggestion
            if (puConfig == null) {
                // Leave Envers unconfigured, but still activate it.
                return;
            }
            if (puConfig.active.isPresent() && !puConfig.active.get()) {
                propertyCollector.accept(EnversService.INTEGRATION_ENABLED, "false");
                // Do not process other properties: Hibernate Envers is inactive anyway.
                return;
            }
```

**Review by yrodiere — CHANGES_REQUESTED** [review]
Thanks! I have a few comments, see below. The nullcheck deserves a second look IMO.

File: extensions/hibernate-envers/deployment/src/test/java/io/quarkus/hibernate/orm/envers/EnversConfigurationPerPUTest.java (hunk 2 above)
Comment (tmihalac) [reply extensions/hibernate-envers/deployment/src/test/java/io/quarkus/hibernate/orm/envers/EnversConfigurationPerPUTest.java:76]: You are right

**Review by yrodiere — APPROVED** [review]
Thanks! LGTM, let's merge once CI passes unless there's an objection.

**Comment by Sanne** [comment]
Awesome, thanks a lot!
